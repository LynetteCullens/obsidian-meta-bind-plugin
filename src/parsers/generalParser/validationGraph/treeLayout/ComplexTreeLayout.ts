import { TL_Element, TL_Literal, TL_Loop, TL_LoopBound, TL_Or } from './TreeLayout';
import { PT_Element_Type } from '../../ParsingTree';
import { AbstractToken } from '../../ParsingUtils';

export abstract class Abstract_TL_C<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> {
	public key: Key | undefined;

	protected constructor(key: Key | undefined) {
		this.key = key;
	}

	abstract toTL(): TL_Element<TokenType, Token, Key>;
}

export class TL_C_Optional<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> extends Abstract_TL_C<TokenType, Token, Key> {
	readonly option: ComplexTreeLayout<TokenType, Token, Key>;

	constructor(option: ComplexTreeLayout<TokenType, Token, Key>, key?: Key | undefined) {
		super(key);

		this.option = option;
	}

	public toTL(): TL_Element<TokenType, Token, Key> {
		return new TL_Or([[], this.option.map(x => x.toTL())], this.key);
	}
}

export class TL_C_Loop<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> extends Abstract_TL_C<TokenType, Token, Key> {
	readonly loop: ComplexTreeLayout<TokenType, Token, Key>;
	readonly min: number;
	readonly max: number;

	constructor(loop: ComplexTreeLayout<TokenType, Token, Key>, min: number = -1, max: number = -1, key?: Key | undefined) {
		super(key);

		this.loop = loop;
		this.min = min;
		this.max = max;
	}

	public toTL(): TL_Element<TokenType, Token, Key> {
		return new TL_Loop(
			this.loop.map(x => x.toTL()),
			new TL_LoopBound(this.min, this.max),
			this.key
		);
	}
}

export class TL_C_Enumeration<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> extends Abstract_TL_C<
	TokenType,
	Token,
	Key
> {
	readonly loop: ComplexTreeLayout<TokenType, Token, Key>;
	readonly separator: ComplexTreeLayout<TokenType, Token, Key>;

	constructor(loop: ComplexTreeLayout<TokenType, Token, Key>, separator: ComplexTreeLayout<TokenType, Token, Key>, key?: Key | undefined) {
		super(key);

		this.loop = loop;
		this.separator = separator;
	}

	public toTL(): TL_Element<TokenType, Token, Key> {
		// TODO: test the loop bound 1, 1
		// the loop bound 1, 1 is done because of the key
		return new TL_Or([
			[],
			[
				new TL_Loop(
					this.loop.map(x => x.toTL()),
					new TL_LoopBound(1, 1),
					this.key
				),
			],
			[
				new TL_Loop([...this.loop.map(x => x.toTL()), ...this.separator.map(x => x.toTL())], new TL_LoopBound(1, -1), this.key),
				new TL_Loop(
					this.loop.map(x => x.toTL()),
					new TL_LoopBound(1, 1),
					this.key
				),
			],
		]);

		// return new TL_Loop(
		// 	this.loop.map(x => x.toTL()),
		// 	new TL_LoopBound(this.min, this.max)
		// );
	}
}

export class TL_C_Or<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> extends Abstract_TL_C<TokenType, Token, Key> {
	readonly options: ComplexTreeLayout<TokenType, Token, Key>[];

	constructor(options: ComplexTreeLayout<TokenType, Token, Key>[]) {
		super(undefined);

		this.options = options;
	}

	public toTL(): TL_Element<TokenType, Token, Key> {
		return new TL_Or(this.options.map(x => x.map(y => y.toTL())));
	}
}

export class TL_C_Literal<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> extends Abstract_TL_C<TokenType, Token, Key> {
	readonly astType: PT_Element_Type;
	readonly tokenType: TokenType | undefined;
	readonly literalContent: string | undefined;

	constructor(astType: PT_Element_Type, tokenType?: TokenType | undefined, literalContent?: string | undefined, key?: Key | undefined) {
		super(key);
		this.astType = astType;
		this.tokenType = tokenType;
		this.literalContent = literalContent;
	}

	public toTL(): TL_Element<TokenType, Token, Key> {
		return new TL_Literal(this.astType, this.tokenType, this.literalContent, this.key);
	}
}

export type TL_C_Element<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> =
	| TL_C_Literal<TokenType, Token, Key>
	| TL_C_Loop<TokenType, Token, Key>
	| TL_C_Or<TokenType, Token, Key>
	| TL_C_Optional<TokenType, Token, Key>
	| TL_C_Enumeration<TokenType, Token, Key>;
export type ComplexTreeLayout<TokenType extends string, Token extends AbstractToken<TokenType>, Key extends string> = TL_C_Element<TokenType, Token, Key>[];
