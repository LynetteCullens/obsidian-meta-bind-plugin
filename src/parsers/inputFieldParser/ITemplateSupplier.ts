import { type UnvalidatedInputFieldDeclaration } from './InputFieldDeclaration';

export interface TemplateSupplierTemplate<T> {
	readonly name: string;
	readonly template: Readonly<T>;
}

export interface ITemplateSupplier<T> {
	getTemplate(templateName: string): Readonly<T> | undefined;
}

export class EmptyTemplateSupplier implements ITemplateSupplier<UnvalidatedInputFieldDeclaration> {
	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	public getTemplate(templateName: string): Readonly<UnvalidatedInputFieldDeclaration> | undefined {
		return undefined;
	}
}
