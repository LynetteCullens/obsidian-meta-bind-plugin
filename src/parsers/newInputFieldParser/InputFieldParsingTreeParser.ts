import { ParsingError } from './ParsingError';
import { ErrorLevel } from '../../utils/errors/MetaBindErrors';
import { ParsingTree, PT_Closure, PT_Element, PT_Element_Type, PT_Literal } from './ParsingTree';
import { Closure } from './ParsingUtils';
import { InputFieldClosures, InputFieldToken, InputFieldTokenType } from './InputFieldTokenizer';

export class InputFieldParsingTreeParser {
	private readonly tokens: InputFieldToken[];
	private readonly closureStack: Closure<InputFieldTokenType>[];
	private readonly parsingTree: ParsingTree;
	private position: number;

	constructor(str: string, tokens: InputFieldToken[]) {
		this.tokens = tokens;
		this.position = 0;
		this.closureStack = [];
		this.parsingTree = new ParsingTree(str, tokens);
	}

	public parse(): ParsingTree {
		while (this.getCurrentToken().type !== InputFieldTokenType.EOF) {
			const astel = this.parseCurrentToken();
			this.parsingTree.children.push(astel);

			if (this.position >= this.tokens.length) {
				throw new Error('index to big');
			}
		}

		return this.parsingTree;
	}

	private parseCurrentToken(): PT_Element {
		const token = this.getCurrentToken();

		this.throwOnInvalidToken();

		const ptLiteral = new PT_Literal(token, this.parsingTree.str);

		for (const closure of InputFieldClosures) {
			const ptClosure = this.parseClosure(ptLiteral, closure);
			if (ptClosure) {
				return ptClosure;
			}
		}

		// move the position to the next token
		this.position += 1;

		return ptLiteral;
	}

	private parseClosure(openingLiteral: PT_Literal, closure: Closure<InputFieldTokenType>): PT_Element | undefined {
		if (openingLiteral.token.type !== closure.openingTokenType) {
			return undefined;
		}

		this.closureStack.push(closure);

		let closingLiteral: PT_Literal | undefined;
		const children: PT_Element[] = [];

		// skip the opening token
		this.position += 1;

		while (this.getCurrentToken().type !== InputFieldTokenType.EOF) {
			const nestedRes = this.parseCurrentToken();

			if (nestedRes.type === PT_Element_Type.LITERAL && (nestedRes as PT_Literal).token.type === closure.closingTokenType) {
				closingLiteral = nestedRes as PT_Literal;
				break;
			} else {
				children.push(nestedRes);
			}
		}

		if (!closingLiteral) {
			// ERROR
			throw new ParsingError(ErrorLevel.ERROR, 'failed to parse', 'closure was not closed', {}, this.parsingTree.str, openingLiteral.token, 'PT Parser');
		}

		this.closureStack.pop();

		return new PT_Closure(this.parsingTree.str, openingLiteral, closingLiteral, children);
	}

	private getCurrentToken(): InputFieldToken {
		return this.tokens[this.position];
	}

	throwOnInvalidToken(): void {
		const token = this.getCurrentToken();

		// check for closure closing tokens that do not actually close a closure
		const currentClosure = this.closureStack.length > 0 ? this.closureStack[this.closureStack.length - 1] : undefined;

		for (const closure of InputFieldClosures) {
			// if the closure is the current token
			if (
				currentClosure &&
				closure.openingTokenType === currentClosure.openingTokenType &&
				closure.closingTokenType === currentClosure.closingTokenType
			) {
				continue;
			}

			// if the current token is a closing token of a closure that is not the active closure
			if (token.type === closure.closingTokenType) {
				throw new ParsingError(
					ErrorLevel.ERROR,
					'failed to parse',
					'Encountered invalid token. Active closure is not the closure that this token closes.',
					{},
					this.parsingTree.str,
					token,
					'PT Parser'
				);
			}
		}
	}
}
