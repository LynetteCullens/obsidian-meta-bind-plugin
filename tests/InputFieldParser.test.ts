import { InputFieldParsingTreeParser } from '../src/parsers/newInputFieldParser/InputFieldParsingTreeParser';
import { InputFieldToken, InputFieldTokenizer, InputFieldTokenType } from '../src/parsers/newInputFieldParser/InputFieldTokenizer';
import { ValidationGraph } from '../src/parsers/newInputFieldParser/validationGraph/ValidationGraph';
import { ParsingTree, PT_Element_Type } from '../src/parsers/newInputFieldParser/ParsingTree';
import { TL_C_Literal, TL_C_Or } from '../src/parsers/newInputFieldParser/validationGraph/treeLayout/ComplexTreeLayout';
import { UnvalidatedInputFieldDeclaration } from '../src/parsers/newInputFieldParser/InputFieldDeclarationValidator';
import { InputFieldStructureParser } from '../src/parsers/newInputFieldParser/InputFieldStructureParser';
import { ErrorCollection } from '../src/utils/errors/ErrorCollection';
import { InputFieldValidationGraphSupplier } from '../src/parsers/newInputFieldParser/validationGraph/InputFieldValidationGraphSupplier';
import { EmptyTemplateSupplier } from '../src/parsers/newInputFieldParser/ITemplateSupplier';

describe('tokenizer tests', function () {
	test('tokenize all simple tokens', () => {
		const input = '()[]:#,.';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		expect(tokens).toMatchSnapshot();
	});

	test('tokenize simple word', () => {
		const input = 'hello world';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		expect(tokens).toMatchSnapshot();
	});

	test('tokenize simple mix of tokens and words', () => {
		const input = 'hello world(this is a message to all people)';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		expect(tokens).toMatchSnapshot();
	});

	test('tokenize simple quotes', () => {
		const input = "'hello world(this is a message to all people)'";
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		expect(tokens).toMatchSnapshot();
	});

	test('tokenize complex quotes', () => {
		const input = "hello world('this is a message to all people []' test [] ':')";
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		expect(tokens).toMatchSnapshot();
	});
});

describe('ast parser tests', function () {
	test('simple closure test', () => {
		const input = 'hello world(this is a message to all people)';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(input, tokens);
		const ast = astParser.parse();
		expect(ast).toMatchSnapshot();
	});

	test('complex closure test 2', () => {
		const input = 'hello world[this is a message to all people]test(a)b';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(input, tokens);
		const ast = astParser.parse();
		console.log(ast.toDebugString());
		expect(ast).toMatchSnapshot();
	});

	test('unclosed closure test', () => {
		const input = 'hello world[this is a message to all (people]test)b';
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(input, tokens);
		expect(() => astParser.parse()).toThrowErrorMatchingSnapshot();
	});

	test('simple input field declaration 1', () => {
		const input = "INPUT[text:nested['object']]";
		const tokenizer = new InputFieldTokenizer(input);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(input, tokens);
		const ast = astParser.parse();
		expect(ast).toMatchSnapshot();
	});
});

describe('tree validation tests', function () {
	function createParsingTree(str: string): ParsingTree<InputFieldTokenType, InputFieldToken> {
		const tokenizer = new InputFieldTokenizer(str);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(str, tokens);
		return astParser.parse();
	}

	test('validation graph test 1', () => {
		const graph = new ValidationGraph([
			new TL_C_Literal(PT_Element_Type.CLOSURE),
			new TL_C_Literal(PT_Element_Type.LITERAL),
			new TL_C_Literal(PT_Element_Type.CLOSURE),
		]);

		expect(graph.validateParsingTree(createParsingTree('()a[]'))).toEqual(true);

		expect(graph.validateParsingTree(createParsingTree('()a[]b'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()[]'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree(''))).toEqual(false);
	});

	test('validation graph test 2', () => {
		const graph = new ValidationGraph([
			new TL_C_Literal(PT_Element_Type.CLOSURE),
			new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.LITERAL)]]),
			new TL_C_Literal(PT_Element_Type.CLOSURE),
		]);

		expect(graph.validateParsingTree(createParsingTree('()a[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()[]'))).toEqual(true);

		expect(graph.validateParsingTree(createParsingTree('()a[]b'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree(''))).toEqual(false);
	});

	test('validation graph test 3', () => {
		const graph = new ValidationGraph([
			new TL_C_Literal(PT_Element_Type.CLOSURE),
			new TL_C_Or([[new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.LITERAL)]])], [new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.CLOSURE)]])]]),
			new TL_C_Literal(PT_Element_Type.CLOSURE),
		]);

		graph.optimize();

		expect(graph.validateParsingTree(createParsingTree('()a[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()()[]'))).toEqual(true);

		expect(graph.validateParsingTree(createParsingTree('()a[]b'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree(''))).toEqual(false);
	});

	test('validation graph test 4', () => {
		const graph = new ValidationGraph([
			new TL_C_Literal(PT_Element_Type.CLOSURE),
			new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.LITERAL)], [new TL_C_Literal(PT_Element_Type.CLOSURE)]]),
			new TL_C_Literal(PT_Element_Type.CLOSURE),
		]);

		expect(graph.validateParsingTree(createParsingTree('()a[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()()[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()[]'))).toEqual(true);

		expect(graph.validateParsingTree(createParsingTree('()a[]b'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree(''))).toEqual(false);
	});

	test('validation graph test 5', () => {
		const graph = new ValidationGraph([
			new TL_C_Literal(PT_Element_Type.CLOSURE),
			new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.LITERAL)]]),
			new TL_C_Or([[], [new TL_C_Literal(PT_Element_Type.CLOSURE)]]),
			new TL_C_Literal(PT_Element_Type.CLOSURE),
		]);

		expect(graph.validateParsingTree(createParsingTree('()a()[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()a[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()()[]'))).toEqual(true);
		expect(graph.validateParsingTree(createParsingTree('()[]'))).toEqual(true);

		expect(graph.validateParsingTree(createParsingTree('()a[]b'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree('()'))).toEqual(false);
		expect(graph.validateParsingTree(createParsingTree(''))).toEqual(false);
	});
});

describe('structure parser test', () => {
	const graphSupplier = new InputFieldValidationGraphSupplier();
	const emptyTemplateSupplier = new EmptyTemplateSupplier();

	function createStructureParser(str: string): InputFieldStructureParser {
		const tokenizer = new InputFieldTokenizer(str);
		const tokens = tokenizer.getTokens();
		const astParser = new InputFieldParsingTreeParser(str, tokens);
		const parsingTree = astParser.parse();
		const errorCollection = new ErrorCollection('structure parser test');

		return new InputFieldStructureParser(emptyTemplateSupplier, graphSupplier, str, tokens, parsingTree, errorCollection);
	}
});
