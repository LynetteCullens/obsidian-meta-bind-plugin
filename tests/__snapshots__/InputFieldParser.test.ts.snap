// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast parser tests complex closure test 2 1`] = `
ParsingTree {
  "children": [
    PT_Literal {
      "str": "hello world[this is a message to all people]test(a)b",
      "token": {
        "literal": "hello world",
        "range": {
          "from": 0,
          "to": 10,
        },
        "type": "WORD",
      },
      "type": "LITERAL",
    },
    PT_Closure {
      "children": [
        PT_Literal {
          "str": "hello world[this is a message to all people]test(a)b",
          "token": {
            "literal": "this is a message to all people",
            "range": {
              "from": 12,
              "to": 42,
            },
            "type": "WORD",
          },
          "type": "LITERAL",
        },
      ],
      "endLiteral": PT_Literal {
        "str": "hello world[this is a message to all people]test(a)b",
        "token": {
          "literal": "]",
          "range": {
            "from": 43,
            "to": 43,
          },
          "type": "]",
        },
        "type": "LITERAL",
      },
      "startLiteral": PT_Literal {
        "str": "hello world[this is a message to all people]test(a)b",
        "token": {
          "literal": "[",
          "range": {
            "from": 11,
            "to": 11,
          },
          "type": "[",
        },
        "type": "LITERAL",
      },
      "str": "hello world[this is a message to all people]test(a)b",
      "type": "CLOSURE",
    },
    PT_Literal {
      "str": "hello world[this is a message to all people]test(a)b",
      "token": {
        "literal": "test",
        "range": {
          "from": 44,
          "to": 47,
        },
        "type": "WORD",
      },
      "type": "LITERAL",
    },
    PT_Closure {
      "children": [
        PT_Literal {
          "str": "hello world[this is a message to all people]test(a)b",
          "token": {
            "literal": "a",
            "range": {
              "from": 49,
              "to": 49,
            },
            "type": "WORD",
          },
          "type": "LITERAL",
        },
      ],
      "endLiteral": PT_Literal {
        "str": "hello world[this is a message to all people]test(a)b",
        "token": {
          "literal": ")",
          "range": {
            "from": 50,
            "to": 50,
          },
          "type": ")",
        },
        "type": "LITERAL",
      },
      "startLiteral": PT_Literal {
        "str": "hello world[this is a message to all people]test(a)b",
        "token": {
          "literal": "(",
          "range": {
            "from": 48,
            "to": 48,
          },
          "type": "(",
        },
        "type": "LITERAL",
      },
      "str": "hello world[this is a message to all people]test(a)b",
      "type": "CLOSURE",
    },
    PT_Literal {
      "str": "hello world[this is a message to all people]test(a)b",
      "token": {
        "literal": "b",
        "range": {
          "from": 51,
          "to": 51,
        },
        "type": "WORD",
      },
      "type": "LITERAL",
    },
  ],
  "str": "hello world[this is a message to all people]test(a)b",
  "tokens": [
    {
      "literal": "hello world",
      "range": {
        "from": 0,
        "to": 10,
      },
      "type": "WORD",
    },
    {
      "literal": "[",
      "range": {
        "from": 11,
        "to": 11,
      },
      "type": "[",
    },
    {
      "literal": "this is a message to all people",
      "range": {
        "from": 12,
        "to": 42,
      },
      "type": "WORD",
    },
    {
      "literal": "]",
      "range": {
        "from": 43,
        "to": 43,
      },
      "type": "]",
    },
    {
      "literal": "test",
      "range": {
        "from": 44,
        "to": 47,
      },
      "type": "WORD",
    },
    {
      "literal": "(",
      "range": {
        "from": 48,
        "to": 48,
      },
      "type": "(",
    },
    {
      "literal": "a",
      "range": {
        "from": 49,
        "to": 49,
      },
      "type": "WORD",
    },
    {
      "literal": ")",
      "range": {
        "from": 50,
        "to": 50,
      },
      "type": ")",
    },
    {
      "literal": "b",
      "range": {
        "from": 51,
        "to": 51,
      },
      "type": "WORD",
    },
    {
      "literal": "eof",
      "range": {
        "from": 52,
        "to": 52,
      },
      "type": "EOF",
    },
  ],
  "type": "ROOT",
}
`;

exports[`ast parser tests simple closure test 1`] = `
ParsingTree {
  "children": [
    PT_Literal {
      "str": "hello world(this is a message to all people)",
      "token": {
        "literal": "hello world",
        "range": {
          "from": 0,
          "to": 10,
        },
        "type": "WORD",
      },
      "type": "LITERAL",
    },
    PT_Closure {
      "children": [
        PT_Literal {
          "str": "hello world(this is a message to all people)",
          "token": {
            "literal": "this is a message to all people",
            "range": {
              "from": 12,
              "to": 42,
            },
            "type": "WORD",
          },
          "type": "LITERAL",
        },
      ],
      "endLiteral": PT_Literal {
        "str": "hello world(this is a message to all people)",
        "token": {
          "literal": ")",
          "range": {
            "from": 43,
            "to": 43,
          },
          "type": ")",
        },
        "type": "LITERAL",
      },
      "startLiteral": PT_Literal {
        "str": "hello world(this is a message to all people)",
        "token": {
          "literal": "(",
          "range": {
            "from": 11,
            "to": 11,
          },
          "type": "(",
        },
        "type": "LITERAL",
      },
      "str": "hello world(this is a message to all people)",
      "type": "CLOSURE",
    },
  ],
  "str": "hello world(this is a message to all people)",
  "tokens": [
    {
      "literal": "hello world",
      "range": {
        "from": 0,
        "to": 10,
      },
      "type": "WORD",
    },
    {
      "literal": "(",
      "range": {
        "from": 11,
        "to": 11,
      },
      "type": "(",
    },
    {
      "literal": "this is a message to all people",
      "range": {
        "from": 12,
        "to": 42,
      },
      "type": "WORD",
    },
    {
      "literal": ")",
      "range": {
        "from": 43,
        "to": 43,
      },
      "type": ")",
    },
    {
      "literal": "eof",
      "range": {
        "from": 44,
        "to": 44,
      },
      "type": "EOF",
    },
  ],
  "type": "ROOT",
}
`;

exports[`ast parser tests simple input field declaration 1 1`] = `
ParsingTree {
  "children": [
    PT_Literal {
      "str": "INPUT[text:nested['object']]",
      "token": {
        "literal": "INPUT",
        "range": {
          "from": 0,
          "to": 4,
        },
        "type": "WORD",
      },
      "type": "LITERAL",
    },
    PT_Closure {
      "children": [
        PT_Literal {
          "str": "INPUT[text:nested['object']]",
          "token": {
            "literal": "text",
            "range": {
              "from": 6,
              "to": 9,
            },
            "type": "WORD",
          },
          "type": "LITERAL",
        },
        PT_Literal {
          "str": "INPUT[text:nested['object']]",
          "token": {
            "literal": ":",
            "range": {
              "from": 10,
              "to": 10,
            },
            "type": ":",
          },
          "type": "LITERAL",
        },
        PT_Literal {
          "str": "INPUT[text:nested['object']]",
          "token": {
            "literal": "nested",
            "range": {
              "from": 11,
              "to": 16,
            },
            "type": "WORD",
          },
          "type": "LITERAL",
        },
        PT_Closure {
          "children": [
            PT_Literal {
              "str": "INPUT[text:nested['object']]",
              "token": {
                "literal": "object",
                "range": {
                  "from": 19,
                  "to": 24,
                },
                "type": "WORD",
              },
              "type": "LITERAL",
            },
          ],
          "endLiteral": PT_Literal {
            "str": "INPUT[text:nested['object']]",
            "token": {
              "literal": "]",
              "range": {
                "from": 26,
                "to": 26,
              },
              "type": "]",
            },
            "type": "LITERAL",
          },
          "startLiteral": PT_Literal {
            "str": "INPUT[text:nested['object']]",
            "token": {
              "literal": "[",
              "range": {
                "from": 17,
                "to": 17,
              },
              "type": "[",
            },
            "type": "LITERAL",
          },
          "str": "INPUT[text:nested['object']]",
          "type": "CLOSURE",
        },
      ],
      "endLiteral": PT_Literal {
        "str": "INPUT[text:nested['object']]",
        "token": {
          "literal": "]",
          "range": {
            "from": 27,
            "to": 27,
          },
          "type": "]",
        },
        "type": "LITERAL",
      },
      "startLiteral": PT_Literal {
        "str": "INPUT[text:nested['object']]",
        "token": {
          "literal": "[",
          "range": {
            "from": 5,
            "to": 5,
          },
          "type": "[",
        },
        "type": "LITERAL",
      },
      "str": "INPUT[text:nested['object']]",
      "type": "CLOSURE",
    },
  ],
  "str": "INPUT[text:nested['object']]",
  "tokens": [
    {
      "literal": "INPUT",
      "range": {
        "from": 0,
        "to": 4,
      },
      "type": "WORD",
    },
    {
      "literal": "[",
      "range": {
        "from": 5,
        "to": 5,
      },
      "type": "[",
    },
    {
      "literal": "text",
      "range": {
        "from": 6,
        "to": 9,
      },
      "type": "WORD",
    },
    {
      "literal": ":",
      "range": {
        "from": 10,
        "to": 10,
      },
      "type": ":",
    },
    {
      "literal": "nested",
      "range": {
        "from": 11,
        "to": 16,
      },
      "type": "WORD",
    },
    {
      "literal": "[",
      "range": {
        "from": 17,
        "to": 17,
      },
      "type": "[",
    },
    {
      "literal": "object",
      "range": {
        "from": 19,
        "to": 24,
      },
      "type": "WORD",
    },
    {
      "literal": "]",
      "range": {
        "from": 26,
        "to": 26,
      },
      "type": "]",
    },
    {
      "literal": "]",
      "range": {
        "from": 27,
        "to": 27,
      },
      "type": "]",
    },
    {
      "literal": "eof",
      "range": {
        "from": 28,
        "to": 28,
      },
      "type": "EOF",
    },
  ],
  "type": "ROOT",
}
`;

exports[`ast parser tests unclosed closure test 1`] = `
"[MB_PARSING_ERROR] "failed to parse" caused by "Encountered invalid token. Active closure is not the closure that this token closes. You probably forgot a ')' somewhere in font of this token."
hello world[this is a message to all (people]test)b
                                            ^
"
`;

exports[`tokenizer tests tokenize all simple tokens 1`] = `
[
  {
    "literal": "(",
    "range": {
      "from": 0,
      "to": 0,
    },
    "type": "(",
  },
  {
    "literal": ")",
    "range": {
      "from": 1,
      "to": 1,
    },
    "type": ")",
  },
  {
    "literal": "[",
    "range": {
      "from": 2,
      "to": 2,
    },
    "type": "[",
  },
  {
    "literal": "]",
    "range": {
      "from": 3,
      "to": 3,
    },
    "type": "]",
  },
  {
    "literal": ":",
    "range": {
      "from": 4,
      "to": 4,
    },
    "type": ":",
  },
  {
    "literal": "#",
    "range": {
      "from": 5,
      "to": 5,
    },
    "type": "#",
  },
  {
    "literal": ",",
    "range": {
      "from": 6,
      "to": 6,
    },
    "type": ",",
  },
  {
    "literal": ".",
    "range": {
      "from": 7,
      "to": 7,
    },
    "type": ".",
  },
  {
    "literal": "eof",
    "range": {
      "from": 8,
      "to": 8,
    },
    "type": "EOF",
  },
]
`;

exports[`tokenizer tests tokenize complex quotes 1`] = `
[
  {
    "literal": "hello world",
    "range": {
      "from": 0,
      "to": 10,
    },
    "type": "WORD",
  },
  {
    "literal": "(",
    "range": {
      "from": 11,
      "to": 11,
    },
    "type": "(",
  },
  {
    "literal": "this is a message to all people []",
    "range": {
      "from": 13,
      "to": 46,
    },
    "type": "WORD",
  },
  {
    "literal": " test ",
    "range": {
      "from": 48,
      "to": 53,
    },
    "type": "WORD",
  },
  {
    "literal": "[",
    "range": {
      "from": 54,
      "to": 54,
    },
    "type": "[",
  },
  {
    "literal": "]",
    "range": {
      "from": 55,
      "to": 55,
    },
    "type": "]",
  },
  {
    "literal": " ",
    "range": {
      "from": 56,
      "to": 56,
    },
    "type": "WORD",
  },
  {
    "literal": ":",
    "range": {
      "from": 58,
      "to": 58,
    },
    "type": "WORD",
  },
  {
    "literal": ")",
    "range": {
      "from": 60,
      "to": 60,
    },
    "type": ")",
  },
  {
    "literal": "eof",
    "range": {
      "from": 61,
      "to": 61,
    },
    "type": "EOF",
  },
]
`;

exports[`tokenizer tests tokenize simple mix of tokens and words 1`] = `
[
  {
    "literal": "hello world",
    "range": {
      "from": 0,
      "to": 10,
    },
    "type": "WORD",
  },
  {
    "literal": "(",
    "range": {
      "from": 11,
      "to": 11,
    },
    "type": "(",
  },
  {
    "literal": "this is a message to all people",
    "range": {
      "from": 12,
      "to": 42,
    },
    "type": "WORD",
  },
  {
    "literal": ")",
    "range": {
      "from": 43,
      "to": 43,
    },
    "type": ")",
  },
  {
    "literal": "eof",
    "range": {
      "from": 44,
      "to": 44,
    },
    "type": "EOF",
  },
]
`;

exports[`tokenizer tests tokenize simple quotes 1`] = `
[
  {
    "literal": "hello world(this is a message to all people)",
    "range": {
      "from": 1,
      "to": 44,
    },
    "type": "WORD",
  },
  {
    "literal": "eof",
    "range": {
      "from": 46,
      "to": 46,
    },
    "type": "EOF",
  },
]
`;

exports[`tokenizer tests tokenize simple word 1`] = `
[
  {
    "literal": "hello world",
    "range": {
      "from": 0,
      "to": 10,
    },
    "type": "WORD",
  },
  {
    "literal": "eof",
    "range": {
      "from": 11,
      "to": 11,
    },
    "type": "EOF",
  },
]
`;
